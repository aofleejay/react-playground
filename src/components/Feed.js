import React from 'react'
import PropTypes from 'prop-types'
import { Layout, Card } from './common'

const Feed = ({ posts, inputText, onChangeText, onSubmit }) => (
  <Layout>
    <form id="post-form" onSubmit={onSubmit}>
      <input
        id="input-text"
        type="text"
        name="text"
        value={inputText}
        maxLength={10}
        onChange={onChangeText}
      />
      <input
        id="submit-button"
        type="submit"
        value="send"
      />
    </form>
    { posts.map(({ id, title, body }) => (
      <Card
        key={id}
        title={title}
        body={body}
      />
    ))}
  </Layout>
)

Feed.propTypes = {
  posts: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      body: PropTypes.string.isRequired,
    })
  ),
  inputText: PropTypes.string,
  onChangeText: PropTypes.func,
  onSubmit: PropTypes.func,
}

Feed.defaultProps = {
  posts: [],
  inputText: '',
  onChangeText: () => {},
  onSubmit: () => {},
}

export default Feed
