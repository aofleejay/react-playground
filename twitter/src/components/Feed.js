import React, { PureComponent, lazy, Suspense } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Layout, Card } from './common'
import ThemeContext from '../contexts/Theme'
import * as COLORS from '../constants/colors'
const Trending = lazy(() => import('./Trending'))

const Form = styled.form`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px;
  border-bottom: 1px solid ${COLORS.GREY_200};
`
Form.displayName = 'Form'

const TextInput = styled.input.attrs({
  type: 'text',
  required: true,
})`
  flex: 1;
  height: 30px;
  padding: 5px;
  color: ${COLORS.BLACK};
  border: 1px solid ${COLORS.GREY_200};
  border-radius: 5px;
  font-size: 14px;
  outline: none;
`
TextInput.displayName = 'TextInput'

const SubmitButton = styled.input.attrs({
  type: 'submit',
})`
  width: 100px;
  height: 30px;
  padding: 5px 15px;
  margin-left: 10px;
  border: none;
  border-radius: 100px;
  color: ${COLORS.WHITE};
  background-color: ${COLORS.GREY_300};
  font-size: 14px;
  font-weight: bold;
  cursor: pointer;
  outline: none;
`
SubmitButton.displayName = 'SubmitButton'

class Feed extends PureComponent {
  static contextType = ThemeContext

  render() {
    const { posts, inputText, onChangeText, onSubmit } = this.props
    const theme = this.context

    return (
      <Suspense fallback={<div>Loading...</div>}>
        <Layout>
          <Trending />
          <Form
            id="post-form"
            onSubmit={onSubmit}
          >
            <TextInput
              id="input-text"
              name="text"
              placeholder="What's happening?"
              value={inputText}
              maxLength={300}
              onChange={onChangeText}
            />
            <SubmitButton
              id="submit-button"
              value="Send"
            />
          </Form>
          { posts.map(({ id, title, body }) => (
            <Card
              key={id}
              title={title}
              body={body}
            />
          ))}
        </Layout>
      </Suspense>
    )
  }
}

Feed.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string,
    body: PropTypes.string.isRequired,
  })),
  inputText: PropTypes.string,
  onChangeText: PropTypes.func,
  onSubmit: PropTypes.func,
}

Feed.defaultProps = {
  posts: [],
  inputText: '',
  onChangeText: () => {},
  onSubmit: () => {},
}

export default Feed
