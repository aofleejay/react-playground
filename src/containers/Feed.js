import React, { Component } from 'react'
import axios from 'axios'
import Feed from '../components/Feed'

class FeedContainer extends Component {
  state = { posts: [], inputText: '' }

  componentDidMount() {
    this.fetchPosts()
  }

  fetchPosts = () => {
    return axios.get('https://jsonplaceholder.typicode.com/posts')
      .then(posts => { this.setState({ posts: posts.data }) })
      .catch(error => { this.setState({ error }) })
  }

  onChangeText = (event) => {
    this.setState({ inputText: event.target.value })
  }

  onSubmit = (event) => {
    event.preventDefault()
    alert(event.target.text.value)
  }

  render() {
    return (
      <Feed
        posts={this.state.posts}
        inputText={this.state.inputText}
        onChangeText={this.onChangeText}
        onSubmit={this.onSubmit}
      />
    )
  }
}

export default FeedContainer
