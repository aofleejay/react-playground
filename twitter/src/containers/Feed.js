import React, { Component } from 'react'
import axios from 'axios'
import Feed from '../components/Feed'
// import { createPost, postsDatabase } from '../libs/firebase'

const POLLING_INTERVAL = 5000

class FeedContainer extends Component {
  state = { posts: [], inputText: '' }

  componentDidMount() {
    this.fetchPosts()
    this.intervalId = setInterval(this.fetchPosts, POLLING_INTERVAL)
  }

  componentWillUnmount() {
    if (this.intervalId) {
      clearInterval(this.intervalId)
    }
  }

  // subscribePosts = () => {
  //   postsDatabase.on('value', (snapshot) => {
  //     snapshot.forEach((childSnapshot) => {
  //       const childKey = childSnapshot.key;
  //       const childData = childSnapshot.val();
  //       const newPost = { id: childKey, ...childData }
  //       this.setState({ posts: [newPost, ...this.state.posts] });
  //     });
  //   });
  // }

  fetchPosts = () => {
    return axios.get('http://localhost:5000/posts?_sort=id&_order=desc')
      .then(response => { this.setState({ posts: response.data }) })
      .catch(error => { this.setState({ error }) })
  }

  createPost = ({ body, title = 'Anonymous' }) => {
    return axios.post('http://localhost:5000/posts', { body, title })
      .then(response => { this.setState({ posts: [response.data, ...this.state.posts], inputText: '' }) })
      .catch(error => { this.setState({ error, inputText: '' }) })
  }

  onChangeText = (event) => {
    this.setState({ inputText: event.target.value })
  }

  onSubmit = (event) => {
    event.preventDefault()
    this.createPost({ body: event.target.text.value })
  }

  render() {
    return (
      <Feed
        posts={this.state.posts}
        inputText={this.state.inputText}
        onChangeText={this.onChangeText}
        onSubmit={this.onSubmit}
      />
    )
  }
}

export default FeedContainer
