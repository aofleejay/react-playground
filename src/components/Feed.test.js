import React from 'react'
import { shallow } from 'enzyme'
import Feed from './Feed'

jest.mock('../components/common', () => ({
  Layout: 'Layout',
  Card: 'Card',
}))

describe('Test Feed component', () => {
  let props

  beforeEach(() => {
    props = {
      posts: [
        { id: 1, title: 'My Title 1', body: 'My body 1' },
        { id: 2, title: 'My Title 2', body: 'My body 2' },
      ],
      inputText: 'My input text',
      onChangeText: () => {},
      onSubmit: () => {},
    }
  })

  describe('Match its snapshot', () => {
    it('When all props exists', () => {
      const wrapper = shallow(<Feed {...props} />)

      expect(wrapper).toMatchSnapshot()
    })

    it('When no props exists', () => {
      const wrapper = shallow(<Feed />)

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Test submit form', () => {
    afterEach(() => {
      jest.resetAllMocks()
    })

    it('Should call props onSubmit when click submit button', () => {
      props.onSubmit = jest.fn()
      const wrapper = shallow(<Feed {...props} />)
      wrapper.find('#post-form').simulate('submit')
      
      expect(props.onSubmit).toHaveBeenCalledTimes(1)
    })

    it('Should call props onChangeText when typing in text input', () => {
      props.onChangeText = jest.fn()
      const wrapper = shallow(<Feed {...props} />)
      wrapper.find('#input-text').simulate('change')
      
      expect(props.onChangeText).toHaveBeenCalledTimes(1)
    })
  })
})
