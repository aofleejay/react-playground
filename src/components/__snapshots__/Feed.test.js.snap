// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Feed component Match its snapshot When all props exists 1`] = `
<Layout>
  <Form
    id="post-form"
    onSubmit={[Function]}
  >
    <TextInput
      id="input-text"
      maxLength={300}
      name="text"
      onChange={[Function]}
      placeholder="What's happening?"
      type="text"
      value="My input text"
    />
    <SubmitButton
      id="submit-button"
      type="submit"
      value="Send"
    />
  </Form>
  <Card
    body="My body 1"
    key="1"
    title="My Title 1"
  />
  <Card
    body="My body 2"
    key="2"
    title="My Title 2"
  />
</Layout>
`;

exports[`Test Feed component Match its snapshot When no props exists 1`] = `
<Layout>
  <Form
    id="post-form"
    onSubmit={[Function]}
  >
    <TextInput
      id="input-text"
      maxLength={300}
      name="text"
      onChange={[Function]}
      placeholder="What's happening?"
      type="text"
      value=""
    />
    <SubmitButton
      id="submit-button"
      type="submit"
      value="Send"
    />
  </Form>
</Layout>
`;
